### PR TITLE
scheduler: move filter logic to list&watch client

### PR DESCRIPTION
**Pull request by mqliang**

### CODE REVIEW

File: plugin/pkg/scheduler/factory/factory.go [hunk 1]
```diff
@@ -256,8 +256,9 @@ func getNodeConditionPredicate() cache.NodeConditionPredicate {
 
 // Returns a cache.ListWatch that finds all pods that need to be
 // scheduled.
-func (factory *ConfigFactory) createUnassignedPodLW() *cache.ListWatch {
-	return cache.NewListWatchFromClient(factory.Client, "pods", api.NamespaceAll, fields.Set{client.PodHost: ""}.AsSelector())
+func (factory *ConfigFactory) createUnassignedAndRunningPodLW() *cache.ListWatch {
+	selector := parseSelectorOrDie("spec.nodeNam==" + "" + ",status.phase!=" + string(api.PodSucceeded) + ",status.phase!=" + string(api.PodFailed))
```
Comment (davidopp) [inline plugin/pkg/scheduler/factory/factory.go]: typo: nodeNam==

This is probably what is causing
k8s.io/kubernetes/plugin/pkg/scheduler
test to fail. (I didn't look into it.)

**Comment by davidopp** [comment]
Can you explain the motivation for this change?

**Comment by mqliang** [comment]
@davidopp 
every pod scheduling will traversal pods cache to filter the (api.PodSucceeded) and (api.PodFailed) pods, if we move the filter logic to list&watcg client, we could  reduse the filtering time/temporary mem use by pod scheduling.

**Comment by davidopp** [comment]
```
=== RUN TestGenericSchedulerWithExtenders
--- PASS: TestGenericSchedulerWithExtenders (0.00s)
=== RUN TestSelectHost
--- PASS: TestSelectHost (0.00s)
=== RUN TestGenericScheduler
--- FAIL: TestGenericScheduler (0.00s)
	generic_scheduler_test.go:299: Unexpected error: Failed for reason matches and possibly others
	generic_scheduler_test.go:302: Failed : test 9, Expected: 2, Saw: 
=== RUN TestFindFitAllError
--- PASS: TestFindFitAllError (0.00s)
=== RUN TestFindFitSomeError
--- PASS: TestFindFitSomeError (0.00s)
=== RUN TestModeler
--- PASS: TestModeler (0.00s)
=== RUN TestScheduler
--- PASS: TestScheduler (0.01s)
	scheduler_test.go:69: Event(api.ObjectReference{Kind:"Pod", Namespace:"", Name:"foo", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'Scheduled' Successfully assigned foo to machine1
	scheduler_test.go:69: Event(api.ObjectReference{Kind:"Pod", Namespace:"", Name:"foo", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'FailedScheduling' scheduler
	scheduler_test.go:69: Event(api.ObjectReference{Kind:"Pod", Namespace:"", Name:"foo", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'FailedScheduling' Binding rejected: binder
=== RUN TestSchedulerForgetAssumedPodAfterDelete
--- PASS: TestSchedulerForgetAssumedPodAfterDelete (0.00s)
	scheduler_test.go:69: Event(api.ObjectReference{Kind:"Pod", Namespace:"", Name:"foo", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'Scheduled' Successfully assigned foo to machine1
	scheduler_test.go:69: Event(api.ObjectReference{Kind:"Pod", Namespace:"", Name:"bar", UID:"", APIVersion:"v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'Scheduled' Successfully assigned bar to machine1
=== RUN TestSchedulerRateLimitsBinding
--- PASS: TestSchedulerRateLimitsBinding (0.00s)
FAIL
FAIL	k8s.io/kubernetes/plugin/pkg/scheduler	0.105s
```

**Comment by davidopp** [comment]
You can run the unit tests yourself by doing
`hack/test-go.sh`
That way you don't need to wait for the presubmit to run it before you can see the failures.

**Comment by mqliang** [comment]
@davidopp fixed.

**Comment by mqliang** [comment]
@k8s-bot unit test this please

**Comment by davidopp** [comment]
@wojtek-t

File: plugin/pkg/scheduler/factory/factory.go [hunk 2]
```diff
@@ -257,16 +257,17 @@ func getNodeConditionPredicate() cache.NodeConditionPredicate {
 
 // Returns a cache.ListWatch that finds all pods that need to be
 // scheduled.
-func (factory *ConfigFactory) createUnassignedPodLW() *cache.ListWatch {
-	return cache.NewListWatchFromClient(factory.Client, "pods", api.NamespaceAll, fields.Set{client.PodHost: ""}.AsSelector())
+func (factory *ConfigFactory) createUnassignedAndRunningPodLW() *cache.ListWatch {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Can we rename it to "createNonTerminatedPodLW"?
In fact it's mostly what you are checking (phase != succeeded and phase != failed)

File: plugin/pkg/scheduler/factory/factory.go [hunk 3]
```diff
@@ -257,16 +257,17 @@ func getNodeConditionPredicate() cache.NodeConditionPredicate {
...
+	return cache.NewListWatchFromClient(factory.Client, "pods", api.NamespaceAll, selector)
 }
 
 // Returns a cache.ListWatch that finds all pods that are
 // already scheduled.
 // TODO: return a ListerWatcher interface instead?
-func (factory *ConfigFactory) createAssignedPodLW() *cache.ListWatch {
-	return cache.NewListWatchFromClient(factory.Client, "pods", api.NamespaceAll,
-		fields.ParseSelectorOrDie(client.PodHost+"!="))
+func (factory *ConfigFactory) createAssignedAndRunningPodLW() *cache.ListWatch {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: Can we rename it to "createAssignedNonTerminatedPodLW"?

File: plugin/pkg/scheduler/factory/factory.go (hunk 2 above)
Comment (wojtek-t) [inline plugin/pkg/scheduler/factory/factory.go]: For consistency with below, probably "createUnassignedNonTerminatedPodLW"

**Comment by wojtek-t** [comment]
Two minor comments - other than that LGTM

**Comment by mqliang** [comment]
@k8s-bot unit test this please

**Comment by mqliang** [comment]
@wojtek-t Updated. PTAL

**Comment by wojtek-t** [comment]
LGTM - thanks!

**Comment by mqliang** [comment]
@k8s-bot unit test this please

**Pull request merged**